src/textengine.rs

```diff
@@ -1,104 +1,132 @@
-            let end = cap.end() as i32;
-            let annot = Annotation { begin, end };
-            annotations.push(annot);
+            static ref RE: RegexEngine =
+                RegexEngine::new(Regex::new(r"[^.!?]*[.!?]").unwrap(), "sentence".into());
         }
-        cas.insert_annotations("token", annotations);
-        Ok(())
+
+        RE.process(cas)
     }
 }
-impl engine::Engine for RegexEngine {
-    fn process(&self, cas: &mut Cas) -> Result<(), PipelineError> {
-        let mut annotations:Vec<Annotation> =  Vec::new();
-        for cap in self.pattern.find_iter(cas.text.as_str()) {
-            let begin = cap.start() as i32;
-            let end = cap.end() as i32;
-            let annot = Annotation { begin, end };
-            annotations.push(annot);
+
+impl Engine for Tokenizer {
+    fn process(&self, cas: &mut Cas) -> Result<()> {
+        lazy_static! {
+            static ref RE: RegexEngine =
+                RegexEngine::new(Regex::new(r"\w+").unwrap(), "token".into());
         }
-        cas.insert_annotations(self.annotation.as_str(), annotations);
-        Ok(())
+
+        RE.process(cas)
     }
 }
-impl engine::Engine for PrintEngine {
-    fn process(&self, cas: &mut Cas) -> Result<(), PipelineError> {
-        cas.print_annotations(self.annotation.as_str())?;
+
+impl Engine for RegexEngine {
+    fn process(&self, cas: &mut Cas) -> Result<()> {
+        let annotations: Vec<Annotation> = self
+            .pattern
+            .find_iter(cas.text())
+            .map(|cap| Annotation {
+                begin: cap.start(),
+                end: cap.end(),
+            })
+            .collect();
+
+        cas.set_annotations(self.annotation.clone(), annotations);
         Ok(())
     }
 }
-impl engine::Reader for SimpleDocumentReader {
-    fn execute(&mut self, cas: &mut Cas) -> Result<(), PipelineError> {
-        if let Some(pbuf) = self.documents.get(self.document_index as usize) {
-            let path = pbuf.as_path();
-            if !path.is_dir() {
-                cas.text = fs::read_to_string(path)?;
-                self.document_index += 1;
-            }
-        }
-        Ok(())
+
+impl Engine for PrintEngine {
+    fn process(&self, cas: &mut Cas) -> Result<()> {
+        cas.print_annotations(&self.annotation)
     }
-    fn has_next(&mut self) -> bool {
-        self.document_index < self.documents_len
+}
+
+impl Reader for SimpleDocumentReader {
+    fn next_cas(&mut self) -> Option<Result<Cas>> {
+        self.documents.next().map(|path| {
+            std::fs::read_to_string(path)
+                .map(Cas::new)
+                .map_err(PipelineError::IoError)
+        })
     }
-    fn initialize(&mut self) -> Result<(), PipelineError> {
-        for entry in WalkDir::new(self.input_dir.as_str())
+}
+
+impl SimpleDocumentReader {
+    pub fn new(input_dir: &Path) -> Result<Self> {
+        let documents: Vec<_> = WalkDir::new(input_dir)
```

Comment:
Here we filter out bad entries in advance and then collect the paths into a `Vec`. Then we can just store the owning iterator for the vector in the `SimpleDocumentReader`.

Reply:
The whole reader improvements I appreciate it, I didnt think of using the map and filter because I was still getting a grasp of the functional ways in rust (using the closures, they look pretty confusing to me yet).

The RAII part, agree with it, makes much more sense and the code got a lot cleaner with it